File: src/lib/triage/layer1-keywords.js
/**
 * Layer 1: Keyword/regex classifier for home service call triage.
 * Runs synchronously. Returns confident:true when a keyword match is found.
 * Patterns hoisted to module level per Vercel best practice (avoids re-compilation per call).
 *
 * Note: Routine patterns are checked BEFORE emergency patterns to prevent
 * "not urgent" from matching the emergency "urgent" pattern.
 */

const EMERGENCY_PATTERNS = [
  /\b(flooding|flooded|flood)\b/i,
  /\bgas\s*(smell|leak|line)\b/i,
  /\bno\s*(heat|hot\s*water)\b/i,
  /\bsewer\s*(backup|overflow)\b/i,
  /\bpipe\s*(burst|broke|broken)\b/i,
  /\belectrical\s*(fire|sparks?|smoke)\b/i,
  /\bcarbon\s*monoxide\b/i,
  /\b(right\s*now|happening\s*now|emergency|urgent)\b/i,
];

const ROUTINE_PATTERNS = [
  /\b(quote|estimate|next\s*(week|month)|sometime|schedule)\b/i,
  /\b(not\s*urgent|whenever|no\s*rush)\b/i,
];

/**
 * Classify a call transcript using keyword matching.
 *
 * @param {string|null} transcript - The call transcript text.
 * @returns {{ result: 'emergency'|'routine', confident: boolean, matched?: string }}
 */
export function runKeywordClassifier(transcript) {
  if (!transcript || transcript.length < 10) {
    return { result: 'routine', confident: false };
  }

  // Check routine patterns first — prevents "not urgent" from triggering emergency "urgent" match
  for (const pattern of ROUTINE_PATTERNS) {
    if (pattern.test(transcript)) {
      return { result: 'routine', confident: true, matched: pattern.source };
    }
  }

  for (const pattern of EMERGENCY_PATTERNS) {
    if (pattern.test(transcript)) {
      return { result: 'emergency', confident: true, matched: pattern.source };
    }
  }

  return { result: 'routine', confident: false };
}

File: src/lib/triage/layer1_keywords.py
import re

EMERGENCY_PATTERNS = [
    re.compile(r"\b(flooding|flooded|flood)\b", re.IGNORECASE),
    re.compile(r"\bgas\s*(smell|leak|line)\b", re.IGNORECASE),
    re.compile(r"\bno\s*(heat|hot\s*water)\b", re.IGNORECASE),
    re.compile(r"\bsewer\s*(backup|overflow)\b", re.IGNORECASE),
    re.compile(r"\bpipe\s*(burst|broke|broken)\b", re.IGNORECASE),
    re.compile(r"\belectrical\s*(fire|sparks?|smoke)\b", re.IGNORECASE),
    re.compile(r"\bcarbon\s*monoxide\b", re.IGNORECASE),
    re.compile(r"\b(right\s*now|happening\s*now|emergency|urgent)\b", re.IGNORECASE),
]

ROUTINE_PATTERNS = [
    re.compile(r"\b(quote|estimate|next\s*(week|month)|sometime|schedule)\b", re.IGNORECASE),
    re.compile(r"\b(not\s*urgent|whenever|no\s*rush)\b", re.IGNORECASE),
]

# Speaker prefixes as built by post_call.py:
#   f"{'Caller' if t['role'] == 'user' else 'AI'}: {t['content']}"
_CALLER_PREFIX = "Caller:"
_AI_PREFIX = "AI:"


def extract_caller_text(transcript: str | None) -> str:
    """Return only the caller's lines from a 'Caller:'/'AI:'-prefixed transcript.

    The agent's own speech must never drive classification — the prompt makes
    the agent say things like "let me take a look at the schedule", which would
    otherwise confidently match ROUTINE_PATTERNS and downgrade a real emergency.
    Falls back to the full text when the input has no speaker prefixes (raw
    text), and to an empty string when only AI lines are present.
    """
    if not transcript:
        return ""
    lines = transcript.splitlines()
    caller_lines = [line for line in lines if line.startswith(_CALLER_PREFIX)]
    if caller_lines:
        return "\n".join(caller_lines)
    if any(line.startswith(_AI_PREFIX) for line in lines):
        return ""  # agent-only transcript — nothing the caller said to classify
    return transcript


def run_keyword_classifier(transcript: str | None) -> dict:
    caller_text = extract_caller_text(transcript)
    if not caller_text or len(caller_text) < 10:
        return {"result": "routine", "confident": False}

    # Emergency patterns FIRST — an emergency match always wins over any
    # routine match ("gas leak, but no rush" is still an emergency).
    for pattern in EMERGENCY_PATTERNS:
        if pattern.search(caller_text):
            return {"result": "emergency", "confident": True, "matched": pattern.pattern}

    for pattern in ROUTINE_PATTERNS:
        if pattern.search(caller_text):
            return {"result": "routine", "confident": True, "matched": pattern.pattern}

    return {"result": "routine", "confident": False}
